### DISCUSSION TITLE
Как RecentTracks.get работает с дубликатами?

### DISCUSSION THREAD

**Discussion by ilyanikolaich** [Chimildic/goofy#70] [Принцип работы]
В описании функции: 
Включенный параметр: ON_SPOTIFY_RECENT_TRACKS и ON_LASTFM_RECENT_TRACKS.
Возвращаемый массив: Объединение обоих источников с удалением дубликатов треков.
И это несложно проверить - удаление дубликатов треков не сокращает массив треков, полученный RecentTracks.get();

В связи с этим вопрос - а как удаляются дубликаты треков, которые были прослушаны в значительно разное время, например, один трек был слушан 1000 треков назад, и вот заново прослушан сейчас - каким образом удалится его дубль - тот, старый, 1001й или не будет включен в историю только что прослушанный, 1й?

Осознанно задаю вопрос в общем обсуждении, чтобы информация была доступна в дальнейшем.
Спасибо.

**Comment by Chimildic** [comment]
Дубликаты удаляются через `Filter.dedupTracks`. Первый встречающийся трек становится оригиналом. Последующие считаются дубликатом. Поскольку сортировка истории от новых к старым, оригинальным треком будет недавно прослушанный.
Правда это справедливо для одного файла. Когда их два, после объединения идет удаление дубликатов и только затем общая сортировка. То есть если в файле ласт.фм есть более новые, а от спотифай нет, то удалятся из ласта. Решается сменой порядка строчек кода. Сначала сортировка, потом поиск дубликатов. В этом есть практический смысл?

upd: порядок изменен и в основной библиотеке

**Reply by ilyanikolaich** [reply]
>> Решается сменой порядка строчек кода. Сначала сортировка, потом поиск дубликатов.
Действительно, нашел в коде, если поменять местами строчки
```
Filter.dedupTracks(spotifyTracks);
Order.sort(spotifyTracks, 'meta.played_at', 'desc');
```
то по вышеописанной логике работы `Filter.dedupTracks` треки, которые зафиксировались в истории ласта, но не зафиксировались в истории спотифая - сохранятся, как более "свежие".
Вообще, имхо - такая логика мне кажется более правильной, особенно если учесть, что на смартфонах часто для скробблинга используется Pano Scrobbler (в частности у меня с задержкой 10 секунд) - в этом случае как раз треки истории, скипнутые после 10 секунды, попадают в ласт, но не попадают в спотифай.

Если действительно для этого нужно просто поменять местами две строчки кода (и это не должно привести к неясным сбоям в матрице) - могу у себя их поменять, и какое-то время посмотреть, не будет ли чего несуразного.

**Reply by Chimildic** [reply]
Нет, ничего случится не должно. Просто во время написания не задумывался о том, какая версия трека останется.

**Reply by ilyanikolaich** [reply]
У себя в коде поменял местами сортировку с удалением дубликатов. 
Мне кажется, что если явных возражений относительно смены версий записанных в истории треков не появится, то можно сменить код.

**Reply by Chimildic** [reply]
Порядок изменен и в основном коде
